pkg/networkservice/common/begin/server.go

```diff
@@ -0,0 +1,102 @@
+// Copyright (c) 2021 Cisco and/or its affiliates.
+//
+// SPDX-License-Identifier: Apache-2.0
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at:
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package begin
+
+import (
+	"context"
+
+	"github.com/networkservicemesh/api/pkg/api/networkservice"
+	"github.com/pkg/errors"
+	"google.golang.org/protobuf/types/known/emptypb"
+
+	"github.com/networkservicemesh/sdk/pkg/networkservice/core/next"
+)
+
+type beginServer struct {
+	serverMap
+}
+
+// NewServer - creates a new begin chain element
+func NewServer() networkservice.NetworkServiceServer {
+	return &beginServer{}
+}
+
+func (b *beginServer) Request(ctx context.Context, request *networkservice.NetworkServiceRequest) (conn *networkservice.Connection, err error) {
+	// No connection.ID, no service
+	if request.GetConnection().GetId() == "" {
+		return nil, errors.New("request.EventFactory.Id must not be zero valued")
+	}
+	// If some other EventFactory is already in the ctx... we are already running in an executor, and can just execute normally
+	if fromContext(ctx) != nil {
+		return next.Server(ctx).Request(ctx, request)
+	}
+	eventFactoryServer, _ := b.LoadOrStore(request.GetConnection().GetId(),
+		newEventFactoryServer(
+			ctx,
+			func() {
+				b.Delete(request.GetRequestConnection().GetId())
+			},
+		),
+	)
+	<-eventFactoryServer.executor.AsyncExec(func() {
+		currentServerClient, _ := b.LoadOrStore(request.GetConnection().GetId(), eventFactoryServer)
+		if currentServerClient != eventFactoryServer {
+			conn, err = b.Request(ctx, request)
+			return
+		}
+		ctx = withEventFactory(ctx, eventFactoryServer)
+		conn, err = next.Server(ctx).Request(ctx, request)
```

Comment:
Should we use here postpone context? I think it can be useful in case if the goroutine for events is loaded.

Reply:
I'm a a little confused... why would we use postpone.Context(ctx) here?

Reply:
Indeed it is not needed because context can be expired on the client side.